File: dijkstra.java
import java.util.*;
import java.io.*;

class dijkstra {

	static class Node{

		int id;
		int from_id;
		List<Integer> edges_to;
		List<Integer> edges_cost;

		boolean done;
		int cost;

		Node(int id){
			this.id=id;
			this.from_id=-1;
			this.done=false;
			this.cost=MAX;

			edges_to = new ArrayList<Integer>();
			edges_cost = new ArrayList<Integer>();
		}

	}

	public static class compClass implements Comparator<Node> {

		// Sorts so that this returns minus (-1)
		public int compare(Node node1, Node node2){
			if(node1.cost>node2.cost){
				return 1;
			}else if(node1.cost<node2.cost){
				return -1;
			}else{
				return 0;
			}
		}
	}

	public static boolean finish(){
		for(Iterator<Map.Entry<Integer,Node>> it=nodes.entrySet().iterator();
			it.hasNext(); ){
			Map.Entry<Integer,Node> entry=it.next();
			if(entry.getValue().done==false) return false;
		}
		return true;
	}

	public static Map<Integer,Node> nodes = new HashMap<Integer,Node>();
	public static final int MAX=1<<20;

	public static void main(String[] argv) throws java.io.IOException{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		System.out.println("[Dijkstra search]\n");

		int T = Integer.parseInt(br.readLine());

		int f,t,c;
		String str[];
		for(int i=0;i<T;++i){
		    str = br.readLine().split(" ");
			f=Integer.parseInt(str[0]);
			t=Integer.parseInt(str[1]);
			c=Integer.parseInt(str[2]);

			Node node;
			if(nodes.containsKey(f)){
				node=nodes.get(f);
			}else{
				node=new Node(f);
				nodes.put(f,node);
			}
			node.edges_to.add(t);
			node.edges_cost.add(c);

			if(nodes.containsKey(t)){
				node=nodes.get(t);
			}else{
				node=new Node(t);
				nodes.put(t,node);
			}
			node.edges_to.add(f);
			node.edges_cost.add(c);
		}

		int s,g;
		str= br.readLine().split(" ");
		s=Integer.parseInt(str[0]);
		g=Integer.parseInt(str[1]);


		Queue<Node> pq = new PriorityQueue<Node>(1, new compClass());

		Node node;
		node=nodes.get(s);
		node.done=true;
		node.cost=0;

		while(true){

			Node next;
			int cost;
			for(int i=0;i<node.edges_to.size();++i){
				next=nodes.get(node.edges_to.get(i));
				if(next.done) continue;

				cost=node.cost+node.edges_cost.get(i);
				if(cost<next.cost){
					pq.remove(next);

					next.cost=cost;
					next.from_id=node.id;
					pq.add(next);
				}
			}

			node=pq.poll();
			node.done=true;
			if(finish()) break;
		}

		System.out.printf("Min cost is %d ",nodes.get(g).cost);

		node=nodes.get(g);
		List<Integer> min_path_nodes=new ArrayList<Integer>();
		while(true){
			min_path_nodes.add(node.id);
			node=nodes.get(node.from_id);
			if(node==null){
				System.out.print("(");
				Collections.sort(min_path_nodes);
				for(int i=0;i<min_path_nodes.size();++i){
					if(i>0) System.out.printf("->");
					System.out.print(min_path_nodes.get(i));
				}
				System.out.print(")\n");
				break;
			}
		}

	}

}

File: dijkstra.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;
typedef vector<int> vint;
typedef vector<int>::iterator itevi;

#define MAX 1<<20

int DEBUG=0;

class Node{
  public:
	int id;
	vint edges_to;
	vint edges_cost;

	bool done;
	int cost;
	int from_id;

	Node(){};
	Node(int id){
		this->done=false;
		this->id=id;
		this->from_id=-1;
		this->cost=MAX;
	};
};

typedef vector<Node> vNode;
vNode nodes;

Node* get_node(int id){
	for(vNode::iterator it=nodes.begin();\
		it!=nodes.end();++it){
		if(it->id==id){
			return &(*it);
		}
	}

	return NULL;
}

class compClass{
  public:
	bool operator() (Node* a, Node* b){
		// XXX: Sorts so that this returns true
		return a->cost>b->cost;
	}
};

void show_nodes(vNode nodes){
	for(int n=0;n<nodes.size();n++){
		cout<<nodes[n].id<<";";
		cout<<nodes[n].cost<<";";
		cout<<nodes[n].edges_to.size()<<";";
		cout<<nodes[n].edges_cost.size()<<";";
		cout<<endl;
	}
}

void show_queue(vector<Node*> pq){
	Node* node;
	cout<<"pq: ";
	while(!pq.empty()){
		node=pq.back();
		cout<<node->id<<":"<<node->cost<<",";
		pq.pop_back();
	}
	cout<<endl;
}

void show_mid_result(Node* node){
	if(DEBUG==0) return;

	vector<int> min_path_nodes;
	Node* tmp_node=node;
	while(true){
		min_path_nodes.push_back(tmp_node->id);
		if(tmp_node->from_id!=-1){
			tmp_node=get_node(tmp_node->from_id);
		}else{
			sort(min_path_nodes.begin(),min_path_nodes.end());
			itevi ite=min_path_nodes.begin();
			while(ite!=min_path_nodes.end()){
				if(ite!=min_path_nodes.begin()) cout<<"->";
				cout<<*ite;
				ite++;
			}
			break;
		}
	}
	cout<<" "<<"("<<node->cost<<")"<<endl;
}

int main(int argc, char* argv[]){
	printf("[Dijkstra search]\n\n");
	if(argc==2&&1==atoi(argv[1])) DEBUG=1;

	int n_edges;
	cin>>n_edges;

	int f,t,c;
	Node *node,*tnode;
	for(int n=0;n<n_edges;++n){
		cin>>f>>t>>c;

		node=get_node(f);
		if(node==NULL){
			Node newnode(f);
			nodes.push_back(newnode);
			node=&nodes.back();
		}
		node->edges_to.push_back(t);
		node->edges_cost.push_back(c);

		tnode=get_node(t);
		if(tnode==NULL){
			Node newnode(t);
			nodes.push_back(newnode);
			tnode=&nodes.back();
		}
		tnode->edges_to.push_back(f);
		tnode->edges_cost.push_back(c);
	}

	int s,g;
	cin>>s>>g;

	// C++ priority_queue does not have remove method
	// use vector and sort instead
	//priority_queue<Node*, vector<Node*>, compClass> pq;
	vector<Node*> pq;

	node=get_node(s);
	if(node==NULL) exit(100);
	node->cost=0;
	node->done=true;
	show_mid_result(node);

	Node* next;
	int cost;
	while(true){

		for(int n=0;n<node->edges_to.size();++n){

			next = get_node(node->edges_to[n]);
			if(next==NULL) exit(100);
			if(next->done) continue;

			cost = node->cost+node->edges_cost[n];
			if(next->cost > cost){
				next->cost=cost;
				next->from_id=node->id; // The shortest path backforward to start node
				pq.push_back(next);
				sort(pq.begin(),pq.end(),compClass()); // Descending order by cost
				show_mid_result(next);
			}
		}

		//show_nodes(nodes);
		//show_queue(pq);

		// Get the last in descending order
		// Erase() in ascending order takes much longer
		node=pq.back(); pq.pop_back();
		node->done=true;
		if(node->id==g) break;
	}

	cout<<"Min cost is "<<node->cost<<" ";
	vector<int> min_path_nodes;
	while(true){
		min_path_nodes.push_back(node->id);
		node=get_node(node->from_id);
		if(node==NULL){
			sort(min_path_nodes.begin(),min_path_nodes.end());
			itevi ite=min_path_nodes.begin();
			cout<<"(";
			while(ite!=min_path_nodes.end()){
				if(ite!=min_path_nodes.begin()) cout<<"->";
				cout<<*ite;
				ite++;
			}
			cout<<")."<<endl;
			break;
		}
	}

	return 0;

}


